File: 백준/Gold/11054. 가장 긴 바이토닉 부분 수열/가장 긴 바이토닉 부분 수열.cc
#include<iostream>

using namespace std;

int main() {

	int N;
	int arr[1001] = { 0, };
	int dp1[1001] = { 0, };
	int dp2[1001] = { 0, };
	int sol = 0;

	cin >> N;

	for (int i = 1; i <= N; i++) {
		cin >> arr[i];
	}

	dp1[1] = 1;

	for (int i = 2; i <= N; i++) {
		dp1[i] = 1;
		for (int j = 1; j < i; j++) {
			if (arr[i] > arr[j]) {
				if (dp1[i] < dp1[j] + 1) {
					dp1[i] = dp1[j] + 1;
				}
			}
		}
	}

	dp2[N] = 1;

	for (int i = N-1; i >= 1; i--) {
		dp2[i] = 1;
		for (int j = N; j > i; j--) {
			if (arr[j] < arr[i]) {
				if (dp2[j] + 1 > dp2[i]) {
					dp2[i] = dp2[j] + 1;
				}
			}
		}
	}

	for (int i = 1; i <= N; i++) {
		if (sol < dp1[i] + dp2[i]) {
			sol = dp1[i] + dp2[i];
		}
	}

	cout << sol - 1;

	return 0;
}

File: 백준/Gold/11054. 가장 긴 바이토닉 부분 수열/가장 긴 바이토닉 부분 수열.java
import java.io.*;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int n = Integer.parseInt(br.readLine());

        int[] arr = new int[n+1];
        int[] dp1 = new int[n+1];   // 왼쪽부터 가장 긴 증가하는 부분 수열 개수
        int[] dp2 = new int[n+1];   // 오른쪽부터 가장 긴 증가하는 부분 수열 개수

        String[] str = br.readLine().split(" ");

        int sol = 0;

        // 숫자 배열 세팅
        for (int i=0; i<n; i++) {
            arr[i+1] = Integer.parseInt(str[i]);
        }

        dp1[1] = 1;
        for (int i=2; i<=n; i++) {
            dp1[i] = 1;
            for (int j=1; j<i; j++) {
                if (arr[i] > arr[j]) {
                    dp1[i] = Math.max(dp1[i], dp1[j] + 1);
                }
            }
        }

        dp2[n] = 1;
        for (int i=n-1; i>=1; i--) {
            dp2[i] = 1;
            for (int j=n; j>=i; j--) {
                if (arr[i] > arr[j]) {
                    dp2[i] = Math.max(dp2[i], dp2[j] + 1);
                }
            }
        }

        for (int i=1; i<=n; i++) {
            sol = Math.max(sol, dp1[i] + dp2[i]-1);
        }

        bw.write(sol+"");
        bw.flush();
        bw.close();

    }
}